fix: update `math/base/special/tand` to match correct reference implementation

anandkaranubc opened the pull request:
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request:

- Fixes the implementation of `math/base/special/tand` to align with the correct reference implementation.
- Adds improved test fixtures to enhance accuracy and ensure correctness.

## Related Issues

> Does this pull request have any related issues?

This pull request:

- resolves none

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
test: add improved test fixtures

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
fix: update the function according to reference implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/tand/test/fixtures/julia/runner.jl:
```diff
@@ -19,7 +19,7 @@
 
 """
-    gen( domain, name )
+	gen( domain, name )
```
Used tab instead of spaces to prevent EditorConfig lint error here.

anandkaranubc replied:
Reference discussion: https://github.com/stdlib-js/stdlib/pull/5473#discussion_r1976573835

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/tand/test/fixtures/julia/runner.jl:
```diff
@@ -62,9 +62,9 @@ file = @__FILE__;
 # Generate fixture data for negative values:
-x = range( -1.0, stop = -10.0, length = 1000 );
+x = range( -180.0, stop = 0.0, length = 1000 );
 gen( x, "negative.json" );
 
 # Generate fixture data for positive values:
-x = range( 1.0, stop = 10.0, length = 1000 );
+x = range( 0.0, stop = 180.0, length = 1000 );
```
Testing it over the period of `tand` now.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/tand/test/test.js:
```diff
@@ -60,7 +60,7 @@ tape( 'the function computes the tangent of an angle measured in degrees (negati
 		} else {
 			delta = abs( y - expected[i] );
-			tol = 1.4 * EPS * abs( expected[i] );
+			tol = 2.0 * EPS * abs( expected[i] );
```
Had to increase the tolerance by a little amount, from `1.4` to `2.0`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/tand/test/test.native.js:
```diff
@@ -69,7 +69,7 @@ tape( 'the function computes the tangent of an angle measured in degrees (negati
 		} else {
 			delta = abs( y - expected[i] );
-			tol = 1.4 * EPS * abs( expected[i] );
+			tol = 1.8 * EPS * abs( expected[i] );
```
Had to increase the tolerance by a little amount, from `1.4` to `1.8` here in native tests.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/tand/lib/main.js:
```diff
@@ -51,23 +49,7 @@ var isInfinite = require( '@stdlib/assert/is-infinite' );
-	xRad = deg2rad( x );
-
-	return tan( xRad );
+	return sind( x ) / cosd( x );
```
I don't think there is a need to have `isnan` and `isinfinity` checks here as they are already handled by `sind` and `cosd`

anandkaranubc commented:
Reference implementation: https://github.com/JuliaLang/julia/blob/9ef2af9db620644253941e080b5b4986ad0d6091/base/special/trig.jl#L1074

anandkaranubc commented:
I am not sure why two of the native tests are failing here. I can confirm that they pass locally on my end.

It is probably due to lower tolerance thresholds in native tests compared to the JS tests.

anandkaranubc committed:
test: add test case for infinity handling

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@anandkaranubc I'd say go ahead and increase the native test tolerance to `2.0 * EPS` to match the JS tolerances. That should resolve the `test.native.js` CI failures, as well.

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into fix/tand

anandkaranubc committed:
test: use the same tol as JS tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
LGTM!
